src/utils.c

```diff
@@ -412,6 +412,17 @@ int descend(QPTPool_t *ctx, const size_t id, void *args,
         const size_t next_level = work->level + 1;
         const size_t recursion_level = work->recursion_level + 1;
 
+        int d_fd = dirfd(dir);
+        if (d_fd < 0) {
+            /*
+             * We should never get here. glibc's dirfd(3) never return errors, and
+             * Apple's libc only returns an error if the DIR * is invalid, which would
+             * indicate a bug in GUFI.
+             */
+            fprintf(stderr, "BUG: dirfd(3) failed: errno = %d, aborting\n", errno);
+            exit(EXIT_FAILURE);
```

Comment:
Would it make sense to `return` instead of `exit`, so that the rest of the tree is still walked?

Reply:
I like exiting when a bug condition is detected, but ultimately it shouldn't ever happen, so it doesn't really matter that much either way. Updated.